### PR TITLE
ENH: Ease access to ViewerPreferences

### LINKED ISSUE #2105: Support right-to-left ViewerPreferences
## Explanation

I'd like to be able to create a right-to-left (RTL or in PDF, R2L) document. More specifically, I'd like to be able to set a document to open in a two page layout (which I already can do with pypdf) with the lower-numbered page on the _right._

From what I can tell, this would involve adding a `/ViewerPreferences` dictionary to the "catalog" with a `/Direction /R2L` entry.

In the Acrobat UI, this is shown as setting the "Binding" to "Right Edge" (in Properties -> Advanced).

## Code Example

How would your feature be used? (Remove this if it is not applicable.)

```python
from pypdf import PdfReader, PdfWriter

pdf_out = PdfWriter()

pdf_out.page_layout = "/TwoPageRight"
pdf_out.viewer_preferences.direction = "/R2L" # new feature


...  # your new feature in action!
```

### PR DESCRIPTION
**Pull request by pubpub-zz**
closes #2105
still doc and test to be fixed/added

### CODE REVIEW

**Comment by pubpub-zz** [comment]
@MartinThoma 
Can you give me your opinion on this first part

**Comment by pubpub-zz** [comment]
@MartinThoma I've been as far as possible. I have not been able to check if documentation is properly generated

File: tests/test_writer.py [hunk 1]
```diff
@@ -1675,3 +1675,68 @@ def test_damaged_pdf_length_returning_none():
...
+def test_viewerpreferences():
+    """
+    Add Tests for ViewerPreferences
+    https://github.com/py-pdf/pypdf/issues/140#issuecomment-1685380549
+    """
+    url = "https://github.com/py-pdf/pypdf/files/9175966/2015._pb_decode_pg0.pdf"
+    name = "2015._pb_decode_pg0.pdf"
+    reader = PdfReader(BytesIO(get_data_from_url(url, name=name)))
+    v = reader.viewer_preferences
+    assert v.center_window == True  # noqa: E712
```
Comment (MartinThoma) [inline tests/test_writer.py:1690]:
```suggestion
    assert v.center_window is True
```

File: tests/test_writer.py [hunk 2]
```diff
@@ -1675,3 +1675,68 @@ def test_damaged_pdf_length_returning_none():
...
+    writer = PdfWriter(clone_from=reader)
+    v = writer.viewer_preferences
+    assert v.center_window == True  # noqa: E712
```
Comment (MartinThoma) [inline tests/test_writer.py:1693]:
```suggestion
    assert v.center_window is True
```

File: tests/test_writer.py [hunk 3]
```diff
@@ -1675,3 +1675,68 @@ def test_damaged_pdf_length_returning_none():
...
+    v.center_window = False
+    assert (
+        writer._root_object["/ViewerPreferences"]["/CenterWindow"]
+        == False  # noqa: E712
```
Comment (MartinThoma) [inline tests/test_writer.py:1697]:
```suggestion
        is False
```

File: tests/test_writer.py [hunk 4]
```diff
@@ -1675,3 +1675,68 @@ def test_damaged_pdf_length_returning_none():
...
+        writer._root_object["/ViewerPreferences"]["/NonFullScreenPageMode"]
+        == "/UseOutlines"
+    )
+    writer = PdfWriter(clone_from=reader)
+    v = writer.viewer_preferences
+    assert v.center_window == True  # noqa: E712
+    v.center_window = False
+    assert (
+        writer._root_object["/ViewerPreferences"]["/CenterWindow"]
+        == False  # noqa: E712
```
Comment (MartinThoma) [inline tests/test_writer.py:1715]:
```suggestion
        is False
```

File: tests/test_writer.py [hunk 5]
```diff
@@ -1675,3 +1675,68 @@ def test_damaged_pdf_length_returning_none():
...
+    with pytest.raises(ValueError):
+        v.non_fullscreen_pagemode = "/toto"
+    v.non_fullscreen_pagemode = "/UseOutlines"
+    assert (
+        writer._root_object["/ViewerPreferences"]["/NonFullScreenPageMode"]
+        == "/UseOutlines"
+    )
+    writer = PdfWriter(clone_from=reader)
+    v = writer.viewer_preferences
+    assert v.center_window == True  # noqa: E712
```
Comment (MartinThoma) [inline tests/test_writer.py:1711]:
```suggestion
    assert v.center_window is True
```

File: tests/test_writer.py [hunk 6]
```diff
@@ -1675,3 +1675,68 @@ def test_damaged_pdf_length_returning_none():
...
+
+    writer = PdfWriter(clone_from=reader)
+    writer._root_object[NameObject("/ViewerPreferences")] = writer._add_object(
+        writer._root_object["/ViewerPreferences"]
+    )
+    v = writer.viewer_preferences
+    v.center_window = False
+    assert (
+        writer._root_object["/ViewerPreferences"]["/CenterWindow"]
+        == False  # noqa: E712
```
Comment (MartinThoma) [inline tests/test_writer.py:1726]:
```suggestion
        is False
```

**Comment by MartinThoma** [comment]
Looks good to me :-) Nice work :rocket: 

I've seen that there was a draft PR: https://github.com/py-pdf/pypdf/commit/c95b694068d00184dfc8a9b7dceabf7446839ccd.patch

If you used that one, I would mere it with `Co-authored-by: Marc Stober` if that is also fine to you. If you didn't use that one or if you don't think that makes sense, I'll merge it as it is. Just let me know :-)

File: tests/test_writer.py (hunk 1 above)
Comment (pubpub-zz) [reply tests/test_writer.py:1690]: This does not work : the converter forces all the time this but it fails the test

File: tests/test_writer.py (hunk 2 above)
Comment (pubpub-zz) [reply tests/test_writer.py:1693]: same

File: tests/test_writer.py (hunk 3 above)
Comment (pubpub-zz) [reply tests/test_writer.py:1697]: same

File: tests/test_writer.py (hunk 4 above)
Comment (pubpub-zz) [reply tests/test_writer.py:1715]: same

File: tests/test_writer.py (hunk 5 above)
Comment (pubpub-zz) [reply tests/test_writer.py:1711]: same

File: tests/test_writer.py (hunk 6 above)
Comment (pubpub-zz) [reply tests/test_writer.py:1726]: same

**Comment by pubpub-zz** [comment]
> Looks good to me :-) Nice work 🚀
> 
> I've seen that there was a draft PR: https://github.com/py-pdf/pypdf/commit/c95b694068d00184dfc8a9b7dceabf7446839ccd.patch
> 
> If you used that one, I would mere it with `Co-authored-by: Marc Stober` if that is also fine to you. If you didn't use that one or if you don't think that makes sense, I'll merge it as it is. Just let me know :-)

I looked at it but  there is a few errors. However I'm completly open to Co-authored it 😀

File: tests/test_writer.py (hunk 1 above)
Comment (MartinThoma) [reply tests/test_writer.py:1690]: Ah, damn. We might have conflicting rules.

Comment (pubpub-zz) [reply tests/test_writer.py:1690]: for log/other users:
This is not a conflicting rules but just the fact that BooleanObjects are different from bool object therefore "is" not working.  ruff is forcing by default the rewriting  so I've added the #noqa. this is more likely a "bug" in ruff to monitor.

**Pull request merged**